AVX-68863 Terraform changes to support gateway level cipher settings.

avx-pkandoi opened the pull request:
AVX-68863 Terraform changes to support gateway level

Added support for edit flow with v2.5 api.
Added support for edge create and edit.
Aligned terraform name with UI for admins.

avx-pkandoi committed:
AVX-68863 Terraform changes to support gateway level cipher settings.

Added changes to support cipher on gateway creation.

sharathbhatp commented on aviatrix/resource_aviatrix_spoke_gateway.go:
```diff
@@ -632,6 +632,17 @@ func resourceAviatrixSpokeGateway() *schema.Resource {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Phase 2 Perfect Forward Secrecy (PFS) policy. Config Options are enable/disable",
+			},
```
default should be false

avx-pkandoi replied:
done

sharathbhatp commented on aviatrix/resource_aviatrix_spoke_gateway.go:
```diff
@@ -1865,6 +1880,13 @@ func resourceAviatrixSpokeGatewayUpdate(d *schema.ResourceData, meta interface{}
 	}
 
+	if d.HasChange("ph2_encryption_policy") {
+		return fmt.Errorf("updating ph2_encryption_policy is not allowed")
```
editing is supported by backend right?

avx-pkandoi replied:
this is only create_mct_primary_gateway api it doesn't have editing support

sharathbhatp replied:
When we edit the attribute in resource and reapply TF, shouldn't it call edit_gateway API?

avx-pkandoi replied:
its calling new v2.5 api. Added support for same in gateway

avx-pkandoi committed:
AVX-68863 Terraform changes to support gateway level cipher settings.

Updated default value for pfs to disabled.

jpeach commented on aviatrix/resource_aviatrix_spoke_gateway.go:
```diff
@@ -632,6 +632,18 @@ func resourceAviatrixSpokeGateway() *schema.Resource {
+				Default:     "default",
+			},
+			"ph2_pfs_policy": {
+				Type:        schema.TypeString,
```
If you can only enable/disable,. why isn't this TypeBool?

avx-pkandoi replied:
this is how it was decided in future it can have many other field.

jpeach commented on aviatrix/resource_aviatrix_spoke_gateway.go:
```diff
@@ -632,6 +632,18 @@ func resourceAviatrixSpokeGateway() *schema.Resource {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Phase 2 encryption policy. Config options are default/strong.",
+				Default:     "default",
```
Set `.ValidateFunc` to validate the allowed options.

avx-pkandoi replied:
done

jpeach commented on aviatrix/resource_aviatrix_spoke_gateway.go:
```diff
@@ -632,6 +632,18 @@ func resourceAviatrixSpokeGateway() *schema.Resource {
+				Description: "Phase 2 encryption policy. Config options are default/strong.",
+				Default:     "default",
+			},
+			"ph2_pfs_policy": {
```
Suggest calling this a human-readable name, perhaps "ph2_forward_secrecy`.

avx-pkandoi replied:
done changed it to new thing.

jpeach commented on docs/resources/aviatrix_spoke_gateway.md:
```diff
@@ -395,6 +395,9 @@ The following arguments are supported:
 * `private_mode_subnet_zone` - (Optional) Availability Zone of the subnet. Required when Private Mode is enabled on the Controller and `cloud_type` is AWS or AWSGov. Available in Provider version R2.23+.
 * `ha_private_mode_subnet_zone` - (Optional) Availability Zone of the HA subnet. Required when Private Mode is enabled on the Controller and `cloud_type` is AWS or AWSGov with HA. Available in Provider version R2.23+.
 * `enable_ipv6` - (Optional) To enable IPv6 CIDR in Spoke Gateway. Only AWS, Azure, AzureGov and AWSGov are supported.
+* `ph2_encryption_policy` - (Optional) Phase 2 encryption policy. Config options are default/strong.
```
Would the user know what "Phase 2 encryption policy is"? I certainly don't. Suggest that we explain it or put it into context somehow.

avx-pkandoi replied:
made it same as UI

jpeach reviewed (changes requested):
Would be helpful for some better explanatory docs to tell people what these settings are for and how they could use them.

avx-pkandoi committed:
Added edge support.

avx-pkandoi committed:
Merge branch 'master' into punit_avx_68863

avx-pkandoi committed:
Fix lint issue

avx-pkandoi commented:
> Would be helpful for some better explanatory docs to tell people what these settings are for and how they could use them.

https://aviatrix.atlassian.net/wiki/spaces/AVXENG/pages/3183706137/STP+142+Cipher+Upgrade+for+peering+Connections

avx-pkandoi committed:
AVX-68863 Terraform changes to support gateway level

Added support for edit flow with v2.5 api.
Added support for edge.
Aligned terraform name with UI.

rpanzer-aviatrix commented on aviatrix/resource_aviatrix_edge_gateway_selfmanaged.go:
```diff
@@ -340,6 +340,20 @@ func resourceAviatrixEdgeGatewaySelfmanaged() *schema.Resource {
+				Default:      "default",
+			},
+			"tunnel_forward_secrecy": {
+				Type:         schema.TypeString,
```
Nit: Could this also be a boolean? Other options in this resource seem to use true/false to enable and disable options.

avx-pkandoi replied:
Way API is decided and in future it can enhance to have more values.
https://aviatrix.atlassian.net/wiki/spaces/AVXENG/pages/3183706137/STP+142+Cipher+Upgrade+for+peering+Connections

rpanzer-aviatrix commented on goaviatrix/gateway.go:
```diff
@@ -1482,3 +1494,19 @@ func (c *Client) DisableIPv6(gateway *Gateway) error {
 }
+
+// SetGatewayPhase2Policy sets the phase2 encryption and pfs policy for the specified gateway.
+func (c *Client) SetGatewayPhase2Policy(gwName, encPolicy string, pfsPolicy string) error {
```
It would be good to have an acceptance test that exercises this method.

rpanzer-aviatrix replied:
That can likely be moved to a follow up PR after the spoke and transit gateway resources have removed the use of deprecated fields.

avx-pkandoi replied:
it was small once so didn't wrote it.

rpanzer-aviatrix commented on goaviatrix/gateway.go:
```diff
@@ -1482,3 +1494,19 @@ func (c *Client) DisableIPv6(gateway *Gateway) error {
+	}
+	var response GatewayPhase2Policy
+	endpoint := fmt.Sprintf("%s/%s", gatewayPhase2PolicyEndpoint, gwName)
+	err := c.PostAPIContext25(context.Background(), &response, endpoint, request)
```
This should use the context passed to `resourceAviatrixEdgeGatewaySelfmanagedUpdate` are whoever the caller is.
Using context.Background could result in timeouts not being respected.

avx-pkandoi replied:
agree we will implement this in 9.0 with gw group when we using v2.5 api

rpanzer-aviatrix commented on goaviatrix/gateway.go:
```diff
@@ -394,6 +400,12 @@ type FQDNGatwayInfo struct {
+type GatewayPhase2Policy struct {
+	Ph2EncryptionPolicy string `json:"ph2_encryption_policy,omitempty"`
+	Ph2PfsPolicy        string `json:"ph2_pfs_policy,omitempty"`
+	GwGroupName         string `json:"gwgroup_name,omitempty"`
```
Nit: The controller does not really expect the property `gwgroup_name` in the request payload. It's not set here, but looking only at this code it looks like I could send it.

avx-pkandoi replied:
This is written common for set / get api. 
Currently GET API is not used but it needs the Gwgroup name.
So to use common structure i have defined both.

Group name in v2.5 is passed in URL PATH

jpeach replied:
I also don't see a place in the schema where the gateway name is used in a request.

avx-pkandoi replied:
gateway code has changed recently to incorporate gw group name which is not incorporated yet in terraform and maybe will be done in 9.0.

In request gwname i.e gwgroup name is used in path.
in response it came as data.
POST /v2.5/api/ipsec-peering-policy/{gwgroup_name}


ex:
curl --location 'https://54.175.124.198//v2.5/api/ipsec-peering-policy/transit-md-v6' \
--header 'Accept: application/json' \
--header 'Authorization: cid yZ8S3bMRTSlTXdQ2S1si' \
--header 'Content-Type: application/json' \
--data '{
  "ph2_encryption_policy": "strong",
  "ph2_pfs_policy": "disable"
}'
Return 



{
    "gwgroup_name": "transit-md-v6",
    "ph2_encryption_policy": "AES-256-GCM-96, AES-128-GCM-96",
    "ph2_pfs_policy": "21"
}

rpanzer-aviatrix reviewed (commented):
It would be really good to have an acceptance test that exercises the new api call, and use the given context for the API call instead of using a background context.

rpanzer-aviatrix commented on aviatrix/resource_aviatrix_edge_gateway_selfmanaged.go:
```diff
@@ -370,6 +384,8 @@ func marshalEdgeGatewaySelfmanagedInput(d *schema.ResourceData) (*goaviatrix.Edg
 		Longitude:                          d.Get("longitude").(string),
 		RxQueueSize:                        d.Get("rx_queue_size").(string),
+		TunnelEncryptionCipher:             d.Get("tunnel_encryption_cipher").(string),
+		TunnelForwardSecrecy:               d.Get("tunnel_forward_secrecy").(string),
```
Why are there no changes to `resourceAviatrixEdgeGatewaySelfmanagedCreate()` for these 2 properties?
Does that mean they are not set for new gateways?

avx-pkandoi replied:
its updated inside marshalEdgeGatewaySelfmanagedInput() which is common function in resourceAviatrixEdgeGatewaySelfmanagedCreate()

avx-pkandoi committed:
Added test case.

avx-pkandoi commented:
> It would be really good to have an acceptance test that exercises the new api call, and use the given context for the API call instead of using a background context.

done. Added some test.

avx-pkandoi committed:
Fix lint issue.

rpanzer-aviatrix commented:
The test currently fails:
```
    resource_aviatrix_edge_gateway_selfmanaged_test.go:26: Step 1/3 error: Error running apply: exit status 1
        
        Error: failed to edit advertised spoke vpc routes of spoke gateway: "edge-ovk8p": rest API edit_gateway_advertised_cidr Post failed: Spoke edge-ovk8p advertised cidr 10.230.3.0/24 cannot be the same as Edge LAN/VLAN CIDR
        
          with aviatrix_edge_gateway_selfmanaged.test,
          on terraform_plugin_test.tf line 3, in resource "aviatrix_edge_gateway_selfmanaged" "test":
           3: resource "aviatrix_edge_gateway_selfmanaged" "test" {
```

avx-pkandoi committed:
Fixed unit test failing for edge.

avx-pkandoi commented:
> The test currently fails:
> 
> ```
>     resource_aviatrix_edge_gateway_selfmanaged_test.go:26: Step 1/3 error: Error running apply: exit status 1
>         
>         Error: failed to edit advertised spoke vpc routes of spoke gateway: "edge-ovk8p": rest API edit_gateway_advertised_cidr Post failed: Spoke edge-ovk8p advertised cidr 10.230.3.0/24 cannot be the same as Edge LAN/VLAN CIDR
>         
>           with aviatrix_edge_gateway_selfmanaged.test,
>           on terraform_plugin_test.tf line 3, in resource "aviatrix_edge_gateway_selfmanaged" "test":
>            3: resource "aviatrix_edge_gateway_selfmanaged" "test" {
> ```

reply it was due to wrong config by some one else. anyhow i fixed it.

jpeach commented on aviatrix/resource_aviatrix_edge_gateway_selfmanaged.go:
```diff
@@ -340,6 +340,20 @@ func resourceAviatrixEdgeGatewaySelfmanaged() *schema.Resource {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "Stronger encryption algorithms (AES-256-GCM-96) for tunnels. Config options are default/strong.",
+				ValidateFunc: validation.StringInSlice([]string{"default", "strong"}, false),
```
Next time we update ciphers, what would the values be? "stronger", "strongest"?

avx-pkandoi replied:
agree we should pass cipher name but this is designed currently like this.

sharathbhatp commented on aviatrix/resource_aviatrix_edge_gateway_selfmanaged.go:
```diff
@@ -340,6 +340,20 @@ func resourceAviatrixEdgeGatewaySelfmanaged() *schema.Resource {
+			"tunnel_encryption_cipher": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "Stronger encryption algorithms (AES-256-GCM-96) for tunnels. Config options are default/strong.",
```
Can this be
```
Encryption ciphers for gateway peering tunnels. Config options are default (AES-126-GCM-96) or strong (AES-256-GCM-96)
```

avx-pkandoi replied:
done

sharathbhatp commented on aviatrix/resource_aviatrix_edge_gateway_selfmanaged.go:
```diff
@@ -340,6 +340,20 @@ func resourceAviatrixEdgeGatewaySelfmanaged() *schema.Resource {
+			"tunnel_forward_secrecy": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "Perfect Forward Secrecy (PFS) for tunnels. Config Options are enable/disable",
```
Perfect Forward Secrecy (PFS) for gateway peering tunnels. Config Options are enable/disable

avx-pkandoi replied:
done

sharathbhatp reviewed (commented):
minor comments

jpeach reviewed (dismissed):
Seems reasonable overall. I think that calling the setting "stronger" will be an issue eventually, maybe just call it by the cipher name.

avx-pkandoi committed:
Updated the config helper messages.

sairane10 committed:
Merge branch 'master' into punit_avx_68863

jpeach commented on goaviatrix/gateway.go:
```diff
@@ -1482,3 +1494,19 @@ func (c *Client) DisableIPv6(gateway *Gateway) error {
+		Ph2EncryptionPolicy: encPolicy,
+		Ph2PfsPolicy:        pfsPolicy,
+	}
+	var response GatewayPhase2Policy
```
I know that this codebase has some history of using the type type for request and response, but that's a very confusing and fragile practice. I don't think we should perpetuate it.

avx-pkandoi replied:
sure changed it.

avx-pkandoi committed:
Changed tunnel cipher policy to two different structure
for better readability.
